crates/circuit/src/packed_instruction.rs

```diff
@@ -514,10 +514,93 @@ pub struct PackedInstruction {
     /// requires the GIL to even `get` (of course!), which makes implementing `Clone` hard for us.
     /// We can revisit once we're on PyO3 0.22+ and have been able to disable its `py-clone`
     /// feature.
-    pub py_op: OnceLock<Py<PyAny>>,
+    py_op: OnceLock<Py<PyAny>>,
 }
 
 impl PackedInstruction {
+    /// Creates a new instance of #[PackedInstruction]
+    pub fn new(
+        op: PackedOperation,
+        qubits: Interned<[Qubit]>,
+        clbits: Interned<[Clbit]>,
+        params: Option<SmallVec<[Param; 3]>>,
+        extra_attrs: ExtraInstructionAttributes,
+    ) -> Self {
+        Self {
+            op,
+            qubits,
+            clbits,
+            params: params.map(Box::new),
+            extra_attrs,
+            #[cfg(feature = "cache_pygates")]
+            py_op: OnceLock::new(),
+        }
+    }
+
+    // Getters
+
+    /// Retrieves an immutable reference to the instruction's underlying operation.
+    pub fn op(&self) -> &PackedOperation {
+        &self.op
+    }
+
+    /// Retrieves an immutable reference to the index under which the interner has stored `qubits`.
+    pub fn qubits(&self) -> Interned<[Qubit]> {
+        self.qubits
+    }
+
+    /// Retrieves an immutable reference to the index under which the interner has stored `clbits`.
+    pub fn clbits(&self) -> Interned<[Clbit]> {
+        self.clbits
+    }
+
+    /// Retrieves an immutable reference to the extra_attributes of this instruction.
+    pub fn extra_attrs(&self) -> &ExtraInstructionAttributes {
+        &self.extra_attrs
+    }
+
+    /// Retrieves the cached py_gate immutably.
+    #[cfg(feature = "cache_pygates")]
+    pub fn py_op(&self) -> &OnceLock<PyObject> {
+        &self.py_op
+    }
+
+    // Setters
+
+    /// Retrieves an immutable reference to the instruction's underlying operation.
+    pub fn op_mut(&mut self) -> &mut PackedOperation {
```

Comment:
Sure, this makes more sense

Reply:
I think making the methods `pub(crate)` is a bit needlessly restrictive.  Anywhere you can get a `&mut PackedInstruction`, you can effectively change any fields you need to, because you can `*instruction = my_new_packed_instruction`, so the actual external access restriction for the circuit needs to be purely managed by when you can get a `&mut PackedInstruction`.

Otherwise, it's just needlessly restrictive - if you can create a new `PackedInstruction` with `pub` visibility, then making these methods `pub(crate)` doesn't actually restrict any capability.